test/cmp_msg_test.c

```diff
@@ -0,0 +1,552 @@
+}
+
+static CMP_MSG_TEST_FIXTURE *set_up(const char *const test_case_name)
+{
+    CMP_MSG_TEST_FIXTURE *fixture;
+
+    if (!TEST_ptr(fixture = OPENSSL_zalloc(sizeof(*fixture))))
+        return NULL;
+    fixture->test_case_name = test_case_name;
+
+    if (!TEST_ptr(fixture->cmp_ctx = OSSL_CMP_CTX_new())
+           || !TEST_true(OSSL_CMP_CTX_set_option(fixture->cmp_ctx,
+                                              OSSL_CMP_OPT_UNPROTECTED_SEND, 1))
+           || !TEST_true(OSSL_CMP_CTX_set1_referenceValue(fixture->cmp_ctx, ref,
+                                                          sizeof(ref)))) {
+        tear_down(fixture);
+        return NULL;
+    }
+    return fixture;
+}
+
+static EVP_PKEY *newkey = NULL;
+static X509 *cert = NULL;
+
+#define EXECUTE_MSG_CREATION_TEST(expr) \
+do { \
+    OSSL_CMP_MSG *msg = NULL; \
+    int good = fixture->expected != 0 ? \
+            TEST_ptr(msg = expr) && TEST_true(valid_asn1_encoding(msg)) : \
+            TEST_ptr_null(msg = expr); \
+ \
+    OSSL_CMP_MSG_free(msg); \
+    return good; \
+} while(0)
+
+/* The following tests call a cmp message creation function.
+ * if fixture->expected != 0:
+ *         returns 1 if the message is created and syntactically correct.
+ * if fixture->expected == 0
+ *         returns 1 if message creation returns NULL                         */
+static int execute_certreq_create_test(CMP_MSG_TEST_FIXTURE *fixture)
+{
+    EXECUTE_MSG_CREATION_TEST(ossl_cmp_certReq_new(fixture->cmp_ctx,
+                                              fixture->bodytype,
+                                              fixture->err_code));
+}
+
+static int execute_errormsg_create_test(CMP_MSG_TEST_FIXTURE *fixture)
+{
+    EXECUTE_MSG_CREATION_TEST(ossl_cmp_error_new(fixture->cmp_ctx, fixture->si,
+                                                 fixture->err_code,
+                                                 NULL/* fixture->free_text */,
+                                                 0));
+}
+
+static int execute_rr_create_test(CMP_MSG_TEST_FIXTURE *fixture)
+{
+    EXECUTE_MSG_CREATION_TEST(ossl_cmp_rr_new(fixture->cmp_ctx));
+}
+
+static int execute_certconf_create_test(CMP_MSG_TEST_FIXTURE *fixture)
+{
+    EXECUTE_MSG_CREATION_TEST(ossl_cmp_certConf_new
+                              (fixture->cmp_ctx, fixture->fail_info, NULL));
+}
+
+static int execute_genm_create_test(CMP_MSG_TEST_FIXTURE *fixture)
+{
+    EXECUTE_MSG_CREATION_TEST(ossl_cmp_genm_new(fixture->cmp_ctx));
+}
+
+static int execute_pollreq_create_test(CMP_MSG_TEST_FIXTURE *fixture)
+{
+    EXECUTE_MSG_CREATION_TEST(ossl_cmp_pollReq_new(fixture->cmp_ctx, 4711));
+}
+
+static int execute_pkimessage_create_test(CMP_MSG_TEST_FIXTURE *fixture)
+{
+    EXECUTE_MSG_CREATION_TEST(ossl_cmp_msg_create
+                              (fixture->cmp_ctx, fixture->bodytype));
+}
+
+#define set1_newPkey(ctx, pkey) \
+    (OSSL_CMP_CTX_set0_newPkey(ctx, 1, pkey) ? EVP_PKEY_up_ref(pkey) : 0)
```

Comment:
Upref first. Only set0 if it succeeds

Reply:
set1_newPkey() is now a static function.